pkg/agent/multicluster/stretched_networkpolicy_controller.go

```diff
@@ -0,0 +1,388 @@
+	if !oldIsPod || !curIsPod {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Pod UpdateFunc received unexpected object", "oldObj", old, "curObj", cur)
+		return
+	}
+	if oldPod != nil && reflect.DeepEqual(oldPod.Labels, curPod.Labels) {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Pod UpdateFunc received a Pod update event, "+
+			"but labels are the same. Skip it", "podName", curPod.Name, "podNamespace", curPod.Namespace)
+		return
+	}
+	s.queue.Add(curPod)
+}
+
+// processPodDelete handles Pod delete events. It deletes the Pod from the
+// labelToPods and podToLabel. After Pod is deleted, its classifier flow will also
+// be deleted. So no need to enqueue this Pod to update its classifier flow.
+func (s *StretchedNetworkPolicyController) processPodDelete(old interface{}) {
+	oldPod, oldIsPod := old.(*v1.Pod)
+	if !oldIsPod {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Pod DeleteFunc received unexpected object", "oldObj", old)
+		return
+	}
+	oldPodRef := getPodReference(oldPod)
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	podLabel := s.podToLabel[oldPodRef]
+	s.deleteLabelToPod(podLabel, oldPod)
+	delete(s.podToLabel, getPodReference(oldPod))
+}
+
+// processNamespaceUpdate handles Namespace update events. It only enqueues all
+// Pods in this Namespace if the Labels of this Namespace has been updated.
+func (s *StretchedNetworkPolicyController) processNamespaceUpdate(old, cur interface{}) {
+	oldNS, oldIsNS := old.(*v1.Namespace)
+	curNS, curIsNS := cur.(*v1.Namespace)
+	if !oldIsNS || !curIsNS {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Namespace UpdateFunc received unexpected object", "oldObj", old, "curObj", cur)
+		return
+	}
+	if reflect.DeepEqual(oldNS.Labels, curNS.Labels) {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Namespace UpdateFunc received a Namespace update event, but labels are the same. Skip it", "Namespace", curNS.Name)
+		return
+	}
+	allPodsInNS, err := s.podLister.Pods(curNS.Name).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Can't list Pods from Namespace", "Namespace", curNS.Name)
+	}
+	for _, pod := range allPodsInNS {
+		s.queue.Add(pod)
+	}
+}
+
+// processLabelIdentityAddOrUpdate handles labelIdentity add or update event. It will
+// 1. Add/Update labelIdentity to labelIdentityCache
+// 2. Enqueue all Pods with this label Identity
+func (s *StretchedNetworkPolicyController) processLabelIdentityAddOrUpdate(obj interface{}) {
+	labelIdentity, isLabelID := obj.(*v1alpha1.LabelIdentity)
+	if !isLabelID {
+		klog.V(2).InfoS("StretchedNetworkPolicyController LabelIdentity Add/UpdateFunc received unexpected object", "obj", obj)
+		return
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.labelIdentityCache[labelIdentity.Spec.Label] = labelIdentity.Spec.ID
+	if podSet, ok := s.labelToPods[labelIdentity.Spec.Label]; ok {
+		for podRef := range podSet {
+			if pod, err := s.podLister.Pods(podRef.Namespace).Get(podRef.Name); err == nil {
```

Comment:
On a high level, I feel this is an expensive operation that is entirely unnecessary: we know the Pod's labels and Pod's namespace labels at this point, so there's no extra information we're getting from the API call at all; plus if there are concurrent Pod update/delete events, they will be queued and handled separately. So IMO, we are simply getting the Pod object here so that we can provide the right typed object in the workqueue. If that is the case, can we modify the workqueue so that the elements are of a different type? Will namespacedName simply work? Or a more lightweight struct which contains info for name, namespace, podLabels and nsLabels?

Reply:
Good one. Changed.